Fixed crash on scanning incomplete id3v2 tags

pokowaka opened the pull request:
- Slight refactor for testability
- Add a set of unit tests
- Fixes crash on bad id3v2 tags
- Contains fix for build break introduced by atv2 support removal (see: e09e81b)

Note, that the "big" change is really just a re-order, as the template requires a specialization before it can be used.

The unit tests exposed a bug in the taglib library.

jmarshallnz commented:
Nice! I'd wait for a team member (currently devcon, so it might be a bit), but suggest you break commit up to:

1. Reorder/template + whitespace changes.
2. Tests.
3. Fixes (separate commits/fix).

By the looks you've already taken care of pushing the fix upstream?

Razzeee commented on xbmc/music/tags/test/TestTagLoaderTagLib.cpp:
```diff
@@ -20,6 +20,155 @@
+};
+
+
+// This test exposes a bug in taglib library (#671) so for now we will not run it for all tag types
```
Thank you for the work.
I can't find #671 can you point me to it or did you mean #670?

pokowaka replied:
I mean https://github.com/taglib/taglib/issues/670 in the taglib library. I've pushed a fix for this issue to the taglib branch.

Razzeee replied:
Might be nice to change that comment then

pokowaka replied:
good point, let me fix that.

pokowaka commented:
@jmarshallnz let me see if I can make the separate commits.

mkortstiege commented:
@pokowaka any chance you can split it up so we can get it merged within this merge window?

pokowaka commented:
Working on it now.

pokowaka committed:
Refactors tagloading class for testability

- Introduce templatized methods to make testing easier
- Reorder methods for templatized methods
- Remove whitespace and commented out code

pokowaka committed:
Adds a set of additional tests to verify tag loading class

- The tests are not green due to various issues in the tag loading
  class.

pokowaka committed:
Contains a series of fixes for tag parsing

- Makes all the unit tests green.
- @Razzeee noticed we are refering to the wrong issue.

pokowaka commented:
@jmarshallnz have made the separation as per your suggestion.  Note that the test is broken with 	4154637, and fixed by a45463f.

Razzeee commented:
Unfortunately the merge window had already closed. 

Can you squash the documentation update worth the 2nd commit where you created the comment?

pokowaka commented:
squashed.

Montellese commented:
@Razzeee: there's no merge window restriction for fixes.

Razzeee commented:
@Montellese 
Yeah, that don't merge cleanups talk in the other PR got me confused.

jenkins build this please

Razzeee commented:
Hey @pokowaka 
I've just seen, that these tests only seem to work under windows :/
See: http://jenkins.kodi.tv/view/Helpers/job/TestMulti-All/391/
Can you have a look again?

pokowaka commented:
Hi,

I don't have access to the results. Odd that the tests fail under linux, as
I develop on a linux box.

Greets,
Erwin.

On Tue, Oct 27, 2015 at 5:28 PM, Razzeee <notifications@github.com> wrote:

> Hey @pokowaka <https://github.com/pokowaka>
> I've just seen, that these tests only seem to work under windows :/
> See: http://jenkins.kodi.tv/view/Helpers/job/TestMulti-All/391/
> Can you have a look again?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8195#issuecomment-151682949>.
>

pokowaka commented:
Investigating.

Greets,
Erwin.

On Wed, Oct 28, 2015 at 3:08 PM, Erwin Jansen <pokowaka@gmail.com> wrote:

> Hi,
>
> I don't have access to the results. Odd that the tests fail under linux,
> as I develop on a linux box.
>
> Greets,
> Erwin.
>
> On Tue, Oct 27, 2015 at 5:28 PM, Razzeee <notifications@github.com> wrote:
>
>> Hey @pokowaka <https://github.com/pokowaka>
>> I've just seen, that these tests only seem to work under windows :/
>> See: http://jenkins.kodi.tv/view/Helpers/job/TestMulti-All/391/
>> Can you have a look again?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/xbmc/xbmc/pull/8195#issuecomment-151682949>.
>>
>
>

pokowaka commented:
Hi,

Likely a taglib issue, clean build on debian with latest taglib:

[----------] Global test environment tear-down
[==========] 583 tests from 91 test cases ran. (9944 ms total)
[  PASSED  ] 583 tests.
Clean shutdown of TestGlobalPattern1
erwinj@mubunti ~/s/xbmc> cat /etc/issue

       master?
Debian GNU/Linux 8 \n \l

Investigating mac 64 build..

Greets,
Erwin.




On Wed, Oct 28, 2015 at 3:13 PM, Erwin Jansen <pokowaka@gmail.com> wrote:

> Investigating.
>
> Greets,
> Erwin.
>
> On Wed, Oct 28, 2015 at 3:08 PM, Erwin Jansen <pokowaka@gmail.com> wrote:
>
>> Hi,
>>
>> I don't have access to the results. Odd that the tests fail under linux,
>> as I develop on a linux box.
>>
>> Greets,
>> Erwin.
>>
>> On Tue, Oct 27, 2015 at 5:28 PM, Razzeee <notifications@github.com>
>> wrote:
>>
>>> Hey @pokowaka <https://github.com/pokowaka>
>>> I've just seen, that these tests only seem to work under windows :/
>>> See: http://jenkins.kodi.tv/view/Helpers/job/TestMulti-All/391/
>>> Can you have a look again?
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/xbmc/xbmc/pull/8195#issuecomment-151682949>.
>>>
>>
>>
>

Razzeee commented:
Weird, I'll try to get some other devs to look at it. As I'm under windows.

pokowaka commented:
Ok,

Created a pull request. It seems that different versions of taglib handle
tag types differently.

Greets,
Erwin.

On Wed, Oct 28, 2015 at 3:33 PM, Erwin Jansen <pokowaka@gmail.com> wrote:

> Hi,
>
> Likely a taglib issue, clean build on debian with latest taglib:
>
> [----------] Global test environment tear-down
> [==========] 583 tests from 91 test cases ran. (9944 ms total)
> [  PASSED  ] 583 tests.
> Clean shutdown of TestGlobalPattern1
> erwinj@mubunti ~/s/xbmc> cat /etc/issue
>
>          master?
> Debian GNU/Linux 8 \n \l
>
> Investigating mac 64 build..
>
> Greets,
> Erwin.
>
>
>
>
> On Wed, Oct 28, 2015 at 3:13 PM, Erwin Jansen <pokowaka@gmail.com> wrote:
>
>> Investigating.
>>
>> Greets,
>> Erwin.
>>
>> On Wed, Oct 28, 2015 at 3:08 PM, Erwin Jansen <pokowaka@gmail.com> wrote:
>>
>>> Hi,
>>>
>>> I don't have access to the results. Odd that the tests fail under linux,
>>> as I develop on a linux box.
>>>
>>> Greets,
>>> Erwin.
>>>
>>> On Tue, Oct 27, 2015 at 5:28 PM, Razzeee <notifications@github.com>
>>> wrote:
>>>
>>>> Hey @pokowaka <https://github.com/pokowaka>
>>>> I've just seen, that these tests only seem to work under windows :/
>>>> See: http://jenkins.kodi.tv/view/Helpers/job/TestMulti-All/391/
>>>> Can you have a look again?
>>>>
>>>> —
>>>> Reply to this email directly or view it on GitHub
>>>> <https://github.com/xbmc/xbmc/pull/8195#issuecomment-151682949>.
>>>>
>>>
>>>
>>
>
